lib/ignore/ignore.go

```diff
@@ -50,6 +58,34 @@ func (p Pattern) String() string {
 	return ret
 }
 
+func (p Pattern) allowsSkippingIgnoredDirs() bool {
+	if p.result.IsIgnored() {
+		return true
+	}
+	if p.pattern[0] != '/' {
+		return false
+	}
+	// Double asterisk everywhere in the path except at the end is bad
+	tocheck := strings.TrimSuffix(p.pattern[1:], "**")
+	if strings.Contains(tocheck, "**") {
+		return false
+	}
+	// Any wildcards (*, [...]) anywhere except for the last path component are bad
+	comps := strings.Split(tocheck, "/")
+	for _, comp := range comps[:len(comps)-1] {
+		if strings.Contains(comp, "*") {
+			return false
+		}
+		// need to check for escapes
+		for i := strings.Index(comp, "["); i != -1; i = strings.Index(comp[i+1:], "[") {
```

Comment:
There are also `?` wildcards and `{bracketed,alternatives}`, should we check for those?

Also escapes are not processed on Windows, so `\[foo]` there is a rooted path with a wildcard expression in it...

Reply:
I really shouldn't handle this globbing stuff manually...  
I switched to using `glob.QuoteMeta` to check for any "special stuff". That has the limitation that it doesn't work with escaped brackets `\[`. So this optimization will not apply to any include patterns with such in them.